pingcap/tidb-operator
# [DNM] cmd: add tkctl check subcommand to check tidb cluster configuration

### What problem does this PR solve? 

We currently don't have any environment checking tools like [tidb-ansible](https://github.com/pingcap/tidb-ansible). The OS kernel parameters, k8s environment and tidb cluster configuration need to be tuned for production deployment. Though we have some [document](https://pingcap.com/docs/v3.0/tidb-in-kubernetes/deploy/prerequisites/) for environment requirement, checking all these is tedious and some settings are tightly related which makes the checking more difficult manually. So we need to provide a similar tool for environment checking.

This PR adds a `check` subcommand for `tkctl` which can verify and check the TiDB cluster configurations for production deployment.

### What is changed and how does it work?

The `tkctl check` subcommand will render the tidb-cluster chart with the provided values. And checks the validation of the PD/TiKV/TiDB configuration file, it also makes some suggestions for some settings like log level, pd max-replicas, tikv block-cache size, tidb prepared-plan-cache, txn-local-latches etc.

*Note:* Though newer versions of PD/TiKV/TiDB support configuration validation, multiple versions of PD/TiDB `config` package can not be used in the same version of tidb-operator due to Go module limitation. Besides TiKV is written in Rust, we can not call the `config` library directly in Go. So I copied and converted the PD/TiKV/TiDB configuration definition, this adds extra maintenance burden to us, we may need to find a better way.

### Check List 

Tests 

 - Manual test
```
tkctl check --chart pingcap/tidb-cluster --values tennix-cluster-values.yaml --chart-version v1.1.0-alpha.1 --cloud aws
```

Code changes

 - Has Go code change

Side effects

No

Related changes

 - Need to cherry-pick to the release-1.1 branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:
 
 ```release-note
add `tkctl check` subcommand for TiDB cluster configuration check
 ```
